pkg/engine/executor/range_aggregation.go

```diff
@@ -0,0 +1,287 @@
+		inputs:     inputs,
+		evaluator:  evaluator,
+		aggregator: &partitionAggregator{entries: make(map[uint64]*partitionEntry)}, // TODO: estimate size during planning
+		opts:       opts,
+	}, nil
+}
+
+// Read reads the next value into its state.
+// It returns an error if reading fails or when the pipeline is exhausted. In this case, the function returns EOF.
+// The implementation must retain the returned error in its state and return it with subsequent Value() calls.
+func (r *RangeAggregationPipeline) Read() error {
+	// if the state already has an error, do not attempt to read.
+	if r.state.err != nil {
+		return r.state.err
+	}
+
+	if r.state.batch != nil {
+		r.state.batch.Release()
+	}
+
+	record, err := r.read()
+	r.state = newState(record, err)
+
+	if err != nil {
+		return fmt.Errorf("run range aggregation: %w", err)
+	}
+	return nil
+}
+
+// TODOs:
+// - Support implicit partitioning by all labels when partitionBy is empty
+// - Use columnar access pattern. Current approach is row-based which does not benefit from the storage format.
+// - Add toggle to return partial results on Read() call instead of returning only after exhausing all inputs.
+func (r *RangeAggregationPipeline) read() (arrow.Record, error) {
+	var (
+		isTSInRange  func(t time.Time) bool
+		tsColumnExpr = &physical.ColumnExpr{
+			Ref: types.ColumnRef{
+				Column: types.ColumnNameBuiltinTimestamp,
+				Type:   types.ColumnTypeBuiltin,
+			},
+		} // timestamp column expression
+
+		// reused on each row read
+		h           = xxhash.New()
+		labelValues = make([]string, len(r.opts.partitionBy))
+	)
+
+	{
+		evalTs := r.opts.endTs
+		earliestTs := r.opts.endTs.Add(-r.opts.rangeInterval)
+		isTSInRange = func(t time.Time) bool {
+			// Aggregate entries that belong in [earliestTs, evalTs)
+			return t.Compare(earliestTs) >= 0 && t.Compare(evalTs) < 0
+		}
+	}
+
+	r.aggregator.Reset() // reset before reading new inputs
+	inputsExhausted := false
+	for !inputsExhausted {
+		inputsExhausted = true
+
+		for _, input := range r.inputs {
+			if err := input.Read(); err != nil {
+				if errors.Is(err, EOF) {
+					continue
+				}
+
+				return nil, err
+			}
+
+			inputsExhausted = false
+			record, _ := input.Value()
+
+			// extract all the columns that are used for partitioning
+			arrays := make([]*array.String, 0, len(r.opts.partitionBy))
+			for _, columnExpr := range r.opts.partitionBy {
+				vec, err := r.evaluator.eval(columnExpr, record)
+				if err != nil {
+					return nil, err
+				}
+
+				if vec.Type() != datatype.String {
+					return nil, fmt.Errorf("unsupported datatype for partitioning %s", vec.Type())
+				}
+
+				arrays = append(arrays, vec.ToArray().(*array.String))
+			}
+
+			// extract timestamp column to check if the entry is in range
+			vec, err := r.evaluator.eval(tsColumnExpr, record)
+			if err != nil {
+				return nil, err
+			}
+			tsCol := vec.ToArray().(*array.Timestamp)
+
+			for row := range int(record.NumRows()) {
```

Comment:
If the records where sorted by timestamp we could use binary search to determine start/end index 😑